### ISSUE TITLE
"Warning: foo.service changed on disk. Run 'systemctl daemon-reload' to reload units." and restarts without prompting

### ISSUE THREAD

**Issue by calimeroteknik** [systemd/systemd#5478]
### Submission type

  - [ ] Bug report
  
- [x] Request for enhancement (RFE)
*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

> 232

*NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!*

### Used distribution

> Arch Linux

### In case of bug report: Expected behaviour you didn't see

> (When stdin is a terminal)
> Having modified a service unit, running `systemctl restart foo.service`
> At the minimum, a prompt/countdown that allows to at least press ^C (in order to run daemon-reload first)
> At best, a prompt that offers to pick one of the following options:
> - restart the old in-memory version of the unit (current default behaviour)
>  → equivalent to the current `systemctl restart foo`
> - stop the unit with what its old version would do, and start it again after reloading it (not necessarily a whole system reload as with daemon-reload)
>  → approximately equivalent to `systemctl stop foo; systemctl daemon-reload; systemctl start foo`
> - reload the unit from disk before restarting it
>  → approximately equivalent to `systemctl daemon-reload; systemctl start foo`
> - exit, don't do anything (through ^C possibly)
> → is currently impossible

### In case of bug report: Unexpected behaviour you saw

> (When stdin is a terminal)
> The service is restarted with its old in-memory version that does not correspond to the unit on-disk
> The administrator is surprised about this behaviour and this can result in unplanned trouble

### In case of bug report: Steps to reproduce the problem

> Change a service `foo` on disk through e.g. /etc/systemd/system/foo.service.d/bar.conf
> Restart the service through `systemctl restart foo`, expecting the changes to take effect

**Comment by poettering** [comment]
Well, I am sorry, but I am not convinced this would really be desirable. Arbitrary commands might want to restart stuff, but it should be under the admin's control when to do a full reload, in particular as that is a relatively expensive operation at the moment.

On Linux we don't have a transactional file system, meaning that individual changes you make are not grouped in any way, and other processes will see your changes piecemeal as you make them. This means if we'd automatically reload configuration at various occasions we might do so at the wrong times, if you make changes on multiple services at once, and really don't want systemd to read the half changed state in the middle of it.

We do automatically reload configuration on all our calls that explicitly change data on disk, such as "systemctl edit", "systemctl enable" and so on. That's because the various changes these tools make semantically may be considered individual transactions, and hence it's appears safe to reload after finish, and that's the deal you make if you invoke these commands (and you can turn it off by passing --no-reload). We do not automatically reload configuration on all calls that just change runtime state, and are at best *consumers* of unit files on disk, but not *producers* of it. Under this category you'll find "systemctl stop", "systemctl start" and "systemctl restart" and similar. That's because if you invoke that you do not imply a transaction, as you don't trigger and change on disk at all, and hence we won't do automatic reloads.

Or to say this differently: reloads really belong to the transaction on disk you make, not to the operation that might read it. Hence: if you make local changes without involving "systemctl enable" or similar, then we assume you know what you do, and you know best when your changes are complete, and you will time the reload properly.

I hope this makes sense. Closing.

**Comment by calimeroteknik** [comment]
I understand the transactional points. I didn't mean to envision changes there, however convenient.

**How about the 'minimum' version:**
When running `systemctl restart foo`, if stdin is a terminal, if the unit changed on disk, prompt before restarting (possibly with a countdown).
This way one can press ^C and run daemon-reload first if it's what was intended.

When systemctl outputs this warning, but restarts the service right away, without leaving you a chance to reload, it can potentially cause trouble since this is often not what the user expected. (e.g. writing to the wrong database…)

It happened to me once more today, hence this thread.
This is something I've read other people complain about as well, with cold sweats and trembling hands.

**Comment by poettering** [comment]
> When running systemctl restart foo, if stdin is a terminal, if the unit changed on disk, prompt before restarting (possibly with a countdown).
> This way one can press ^C and run daemon-reload first if it's what was intended.

I am not convinced this would be optimal. I mean, people invoke scripts on a tty that might call systemctl restart, and which really shouldn't be interactive.

I mean, interactivity is something we should only do if really really necessary (for example, when we really need a password to proceed or so), or if the user explicitly asked for it (think: user gets the mysql SQL prompt when he explicitly asked for it by invoking the "mysql" tool). But becoming interactive sometimes (or slow, i.e. countdown) and other times not, without a very strong reason appears wrong to me, and a potential source of user unhappiness...

**Comment by calimeroteknik** [comment]
Right now it seems we are driven into a corner, without an obvious way to make the behaviour of `systemctl restart` less dangerous without changing much.

I believe that the best scenario would have been this:
When the unit was changed on disk, `systemctl restart foo` warns and exits, but *without restarting the service*. However it offers to re-run with --force or another such option if we really want that.

I only broke a test database today, but if I had systemd on production systems this would have meant downtime at the very least.
And I think that qualifies as unhappiness…
